### PR TITLE
Live: pipeline Loki output

### PR DESCRIPTION
**Pull request by FZambia**
**What this PR does / why we need it**:

Adds the initial support for exporting raw data and frames from Live pipeline to Loki. Data collected into batches and periodically flushed to Loki [push](https://grafana.com/docs/loki/latest/api/#post-lokiapiv1push) endpoint. Similar to what we do in remote write case.

Local Loki can be run like this:

```
docker run -it --rm -v /path/to/loki-config.yml:/mnt/config/loki-config.yml -p 3100:3100 grafana/loki:2.3.0 -config.file=/mnt/config/loki-config.yml
```

<details>
<summary>loki-config.yml</summary>

```yaml
auth_enabled: false

server:
  http_listen_port: 3100

ingester:
  lifecycler:
    address: 127.0.0.1
    ring:
      kvstore:
        store: inmemory
      replication_factor: 1
    final_sleep: 0s
  chunk_idle_period: 5m
  chunk_retain_period: 30s

schema_config:
  configs:
  - from: 2020-05-15
    store: boltdb
    object_store: filesystem
    schema: v11
    index:
      prefix: index_
      period: 168h

storage_config:
  boltdb:
    directory: /tmp/loki/index

  filesystem:
    directory: /tmp/loki/chunks

limits_config:
  enforce_metric_name: false
  reject_old_samples: true
  reject_old_samples_max_age: 168h

```

</details>

And rule can look like this:

```
{
   ...
   "dataOutputs": [
        {"type": "loki", "loki": {"uid": "local-loki"}}
    ]
}
```

where remote write backend configured like this

```
...
    {
      "uid": "local-loki",
      "settings": {
        "endpoint": "http://127.0.0.1:3100/loki/api/v1/push"
      }
    }
```

~~To support remote Loki we should first merge https://github.com/grafana/grafana/pull/40147 which contains some required refactoring for remote write backends.~~

@ryantxu  Added some comments where I need an advice about how to identify logs we send.

Screenshot 2021-11-05 at 18 55 06

### CODE REVIEW

File: pkg/services/live/pipeline/data_output_loki.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	return DataOutputTypeLoki
+}
+
+func (out *LokiDataOutput) OutputData(_ context.Context, vars Vars, data []byte) ([]*ChannelData, error) {
+	if out.lokiWriter.endpoint == "" {
+		logger.Debug("Skip sending to Loki: no url")
+		return nil, nil
+	}
+	err := out.lokiWriter.write(LokiStream{
+		Stream: map[string]string{"channel": vars.Channel},
```
Comment (FZambia) [inline pkg/services/live/pipeline/data_output_loki.go:32]: @ryantxu here is where I am not sure what we should put to identify our log entries.

File: pkg/services/live/pipeline/frame_output_loki.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+	if out.lokiWriter.endpoint == "" {
+		logger.Debug("Skip sending to Loki: no url")
+		return nil, nil
+	}
+	frameJSON, err := data.FrameToJSON(frame, data.IncludeAll)
+	if err != nil {
+		return nil, err
+	}
+	err = out.lokiWriter.write(LokiStream{
+		Stream: map[string]string{"frame": frame.Name, "channel": vars.Channel},
```
Comment (FZambia) [inline pkg/services/live/pipeline/frame_output_loki.go:55]: And also here ^^

File: pkg/services/live/pipeline/data_output_loki.go (hunk 1 above)
Comment (FZambia) [reply pkg/services/live/pipeline/data_output_loki.go:32]: In theory it's possible to define some static labels or dynamic labels in output settings.

**Comment by marefr** [comment]
Maybe we should change so that @grafana/grafana-edge-squad is codeowner of pkg/live?

**Comment by FZambia** [comment]
@marefr we can, for global or tricky changes we can always add backend team to review manually. Could you do this (don't know how)?

File: pkg/services/live/pipeline/config.go [hunk 3]
```diff
@@ -61,12 +63,14 @@ type FrameOutputterConfig struct {
 	ConditionalOutputConfig *ConditionalOutputConfig   `json:"conditional,omitempty"`
 	ThresholdOutputConfig   *ThresholdOutputConfig     `json:"threshold,omitempty"`
 	RemoteWriteOutputConfig *RemoteWriteOutputConfig   `json:"remoteWrite,omitempty"`
+	LokiOutputConfig        *LokiOutputConfig          `json:"loki,omitempty"`
```
Comment (ArturWierzbicki) [inline pkg/services/live/pipeline/config.go:66]: I see `LokiOutputConfig` as both a `Frame-` and a `DataOutputter`, is it about sending the data to Loki in different formats?

**Comment by marefr** [comment]
@FZambia noticed it's already there https://github.com/grafana/grafana/blob/dce851e396672f3c3f7e4916fb383c7665ce42ce/.github/CODEOWNERS#L51

**Comment by marefr** [comment]
But looks like a bug/incorrect configuration:
image

Maybe it should be without slash at the end?  `/pkg/service/live @grafana/grafana-edge-squad`

File: pkg/services/live/pipeline/data_output_loki.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
+package pipeline
+
+import (
+	"context"
+	"time"
+)
+
+// LokiDataOutput passes processing control to the rule defined
```
Comment (FZambia) [inline pkg/services/live/pipeline/data_output_loki.go:8]: oops - copy-paste

File: pkg/services/live/pipeline/data_output_loki.go (hunk 1 above)
Comment (ryantxu) [reply pkg/services/live/pipeline/data_output_loki.go:32]: this is great for now -- soon we should have both static+dynamic labels that could apply here and to loki

File: pkg/services/live/pipeline/config.go [hunk 5]
```diff
@@ -61,12 +63,14 @@ type FrameOutputterConfig struct {
...
 	ChangeLogOutputConfig   *ChangeLogOutputConfig     `json:"changeLog,omitempty"`
 }
 
 type DataOutputterConfig struct {
 	Type                     string                    `json:"type"`
 	RedirectDataOutputConfig *RedirectDataOutputConfig `json:"redirect,omitempty"`
+	LokiOutputConfig         *LokiOutputConfig         `json:"loki,omitempty"`
```
Comment (ryantxu) [inline pkg/services/live/pipeline/config.go:73]: :tada:

**Review by ryantxu — APPROVED** [review]

**Pull request merged**